src/tests/nativeaot/SmokeTests/Preinitialization/Preinitialization.cs

```diff
@@ -834,6 +835,39 @@ public static void Run()
     }
 }
 
+class TestGCInteraction
+{
+    class WithFrozenObjects
+    {
+        internal readonly static string s_someStringLiteral = "Some string literal";
+        internal readonly static object s_someObject = new object();
+    }
+
+    public static void Run()
+    {
+        Assert.IsPreinitialized(typeof(WithFrozenObjects));
+
+        var holder = new object[]
+        {
+            WithFrozenObjects.s_someStringLiteral,
+            WithFrozenObjects.s_someObject,
+        };
+
+        GC.Collect();
+        GC.WaitForPendingFinalizers();
+        GC.Collect();
+
+        Assert.AreSame(holder[0], WithFrozenObjects.s_someStringLiteral);
+        Assert.AreSame(holder[1], WithFrozenObjects.s_someObject);
+
+        // This will try to make a dependent handle for the frozen object
+        lock (WithFrozenObjects.s_someObject)
```

Comment:
Since we have a public DependentHandle type, do we want to use that directly here instead of using a lock?

Reply:
👍 Forgot the API is now public.